tap/tcp_reader.go

```diff
@@ -23,44 +21,44 @@ type tcpReader struct {
 	isClient      bool
 	isOutgoing    bool
 	msgQueue      chan api.TcpReaderDataMsg // Channel of captured reassembled tcp payload
+	buffer        []byte
+	exhaustBuffer bool
 	data          []byte
 	progress      *api.ReadProgress
 	captureTime   time.Time
 	parent        *tcpStream
 	packetsSeen   uint
-	extension     *api.Extension
 	emitter       api.Emitter
 	counterPair   *api.CounterPair
 	reqResMatcher api.RequestResponseMatcher
 	sync.Mutex
 }
 
-func NewTcpReader(msgQueue chan api.TcpReaderDataMsg, progress *api.ReadProgress, ident string, tcpId *api.TcpID, captureTime time.Time, parent *tcpStream, isClient bool, isOutgoing bool, extension *api.Extension, emitter api.Emitter, counterPair *api.CounterPair, reqResMatcher api.RequestResponseMatcher) *tcpReader {
+func NewTcpReader(msgQueue chan api.TcpReaderDataMsg, progress *api.ReadProgress, ident string, tcpId *api.TcpID, captureTime time.Time, parent *tcpStream, isClient bool, isOutgoing bool, emitter api.Emitter) *tcpReader {
 	return &tcpReader{
-		msgQueue:      msgQueue,
-		progress:      progress,
-		ident:         ident,
-		tcpID:         tcpId,
-		captureTime:   captureTime,
-		parent:        parent,
-		isClient:      isClient,
-		isOutgoing:    isOutgoing,
-		extension:     extension,
-		emitter:       emitter,
-		counterPair:   counterPair,
-		reqResMatcher: reqResMatcher,
+		msgQueue:    msgQueue,
+		progress:    progress,
+		ident:       ident,
+		tcpID:       tcpId,
+		captureTime: captureTime,
+		parent:      parent,
+		isClient:    isClient,
+		isOutgoing:  isOutgoing,
+		emitter:     emitter,
 	}
 }
 
 func (reader *tcpReader) run(options *api.TrafficFilteringOptions, wg *sync.WaitGroup) {
 	defer wg.Done()
-	b := bufio.NewReader(reader)
-	err := reader.extension.Dissector.Dissect(b, reader, options)
-	if err != nil {
-		_, err = io.Copy(ioutil.Discard, reader)
-		if err != nil {
-			logger.Log.Errorf("%v", err)
+	for i, extension := range extensions {
+		reader.reqResMatcher = reader.parent.reqResMatchers[i]
+		reader.counterPair = reader.parent.counterPairs[i]
+		b := bufio.NewReader(reader)
+		err := extension.Dissector.Dissect(b, reader, options)
+		if err == nil || reader.isProtocolIdentified() {
```

Comment:
```suggestion
		if reader.isProtocolIdentified() {
```
I get that if there is an error, we still want to process through all of the other extensions because maybe the error is specific to the extension. So we don't break if `err != nil`.
Does `Dissect` guarantee that if the protocol is not identified, then it will return `err == nil`? If not, then this will cause a premature exit from `run`. If yes, then it is sufficient to only check `isProtocolIdentified`.